### PR TITLE
feat: Add Analytics events

### PR DESCRIPTION
**Pull request by omerhabib26**
### Description 

- Add analytics for the following modules
- AppReview, Profile, WhatsNew, Course, Discussion, Logistration, MainDashboard, Discovery, Core
- Restructure the analytics implementation to module level
- Update the event names accordingly
- Update the test cases accordingly

fix: LEARNER-9876

### CODE REVIEW

File: app/src/main/java/org/openedx/app/AppViewModel.kt [hunk 1]
```diff
@@ -51,10 +51,19 @@ class AppViewModel(
...
+
     private fun setUserId() {
         preferencesManager.user?.let {
             analytics.setUserIdForSession(it.id)
         }
     }
 
+    private fun logEvent(event: AppAnalyticEvent, biValue: AppAnalyticValues) {
+        analytics.logEvent(event.event, buildMap {
+            put(AppAnalyticKey.NAME.key, biValue.value)
```
Comment (k1rill) [inline app/src/main/java/org/openedx/app/AppViewModel.kt:66]: nit: please improve formatting for better code reading

File: auth/src/main/java/org/openedx/auth/presentation/logistration/LogistrationViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            LogistrationAnalyticValues.EXPLORE_ALL_COURSES
+        )
+    }
+
+    private fun logEvent(
+        eventName: LogistrationAnalyticEvent,
+        biValue: LogistrationAnalyticValues,
+        params: Map<String, Any?> = emptyMap()
+    ) {
+        analytics.logEvent(eventName.event, buildMap {
```
Comment (k1rill) [inline auth/src/main/java/org/openedx/auth/presentation/logistration/LogistrationViewModel.kt:66]: nit: please improve formatting
```
analytics.logEvent(
    eventName.event,
    buildMap {
        put(LogistrationAnalyticKey.NAME.key, biValue.biValue)
        putAll(params)
    }
)
```

File: auth/src/main/java/org/openedx/auth/presentation/restore/RestorePasswordViewModel.kt [hunk 3]
```diff
@@ -84,4 +87,22 @@ class RestorePasswordViewModel(
...
+    private fun logResetPasswordEvent(success: Boolean) {
+        analytics.logEvent(
+            LogistrationAnalyticEvent.RESET_PASSWORD_SUCCESS.event,
+            buildMap {
+                put(LogistrationAnalyticKey.NAME.key, LogistrationAnalyticValues.RESET_PASSWORD_SUCCESS.biValue)
+                put(LogistrationAnalyticKey.SUCCESS.key, success)
+            }
+        )
+    }
 }
```
Comment (k1rill) [inline auth/src/main/java/org/openedx/auth/presentation/restore/RestorePasswordViewModel.kt:108]: new line

File: auth/src/main/java/org/openedx/auth/presentation/signin/SignInViewModel.kt [hunk 4]
```diff
@@ -158,7 +177,7 @@ class SignInViewModel(
             logger.d { "Social login (${authType.methodName}) success" }
             _uiState.update { it.copy(loginSuccess = true) }
             setUserId()
-            analytics.userLoginEvent(authType.methodName)
+//            analytics.userLoginEvent(authType.methodName)
```
Comment (k1rill) [inline auth/src/main/java/org/openedx/auth/presentation/signin/SignInViewModel.kt:180]: should it be commented?

File: auth/src/main/java/org/openedx/auth/presentation/signin/SignInViewModel.kt [hunk 5]
```diff
@@ -217,4 +236,15 @@ class SignInViewModel(
             }
         }
     }
+
+    private fun logEvent(
+        eventName: LogistrationAnalyticEvent,
+        biValue: LogistrationAnalyticValues,
+        params: Map<String, Any?> = emptyMap(),
+    ) {
+        analytics.logEvent(eventName.event, buildMap {
```
Comment (k1rill) [inline auth/src/main/java/org/openedx/auth/presentation/signin/SignInViewModel.kt:245]: nit: please improve formatting

File: core/src/main/java/org/openedx/core/ui/ComposeCommon.kt [hunk 6]
```diff
@@ -165,7 +165,7 @@ fun Toolbar(
             modifier = Modifier
                 .testTag("txt_toolbar_title")
                 .align(Alignment.CenterVertically)
-                .padding(end = 16.dp),
+                .padding(start = if (canShowBackBtn.not()) 16.dp else 0.dp, end = 16.dp),
```
Comment (k1rill) [inline core/src/main/java/org/openedx/core/ui/ComposeCommon.kt:168]: nit: it's better to use positive logic: if (canShowBackBtn) 0.dp else 16.dp

but you can skip this comment

File: course/src/main/java/org/openedx/course/presentation/container/CourseContainerFragment.kt [hunk 7]
```diff
@@ -205,21 +210,44 @@ class CourseContainerFragment : Fragment(R.layout.fragment_course_container) {
...
+                                CalendarSyncDialogType.OUT_OF_SYNC_DIALOG ->
+                                    viewModel.logCalendarSyncUpdate(false)
+
+                                CalendarSyncDialogType.EVENTS_DIALOG ->
+                                    viewModel.logCalendarSyncedConfirmation(false)
+
+                                CalendarSyncDialogType.LOADING_DIALOG,
+                                CalendarSyncDialogType.PERMISSION_DIALOG,
+                                CalendarSyncDialogType.NONE,
+                                -> {
```
Comment (k1rill) [inline course/src/main/java/org/openedx/course/presentation/container/CourseContainerFragment.kt:247]: nit: change {} to Unit

you can skip this

File: course/src/main/java/org/openedx/course/presentation/section/CourseSectionViewModel.kt [hunk 8]
```diff
@@ -141,10 +146,17 @@ class CourseSectionViewModel(
...
+                CourseAnalyticEvent.UNIT_DETAIL.event,
+                buildMap {
+                    put(CourseAnalyticKey.NAME.key, CourseAnalyticValue.UNIT_DETAIL.biValue)
+                    put(CourseAnalyticKey.COURSE_ID.key, courseId)
+                    put(CourseAnalyticKey.BLOCK_ID.key, blockId)
+                    put(CourseAnalyticKey.CATEGORY.key, CourseAnalyticKey.NAVIGATION.key)
+                })
         }
     }
 }
```
Comment (k1rill) [inline course/src/main/java/org/openedx/course/presentation/section/CourseSectionViewModel.kt:162]: new line

File: course/src/main/java/org/openedx/course/presentation/unit/video/VideoUnitViewModel.kt [hunk 9]
```diff
@@ -128,5 +130,4 @@ open class VideoUnitViewModel(
     }
 
     fun getCurrentVideoTime() = currentVideoTime.value ?: 0
-
 }
```
Comment (k1rill) [inline course/src/main/java/org/openedx/course/presentation/unit/video/VideoUnitViewModel.kt:133]: new line

File: profile/src/main/java/org/openedx/profile/presentation/edit/EditProfileViewModel.kt [hunk 10]
```diff
@@ -128,11 +138,30 @@ class EditProfileViewModel(
...
+                put(
+                    ProfileAnalyticKey.NAME.key,
+                    ProfileAnalyticValue.DELETE_ACCOUNT_CLICKED.biValue
+                )
+            }
+        )
     }
 
+    private fun logProfileEvent(event: String, params: Map<String, Any?>) {
+        analytics.logEvent(event, buildMap {
```
Comment (k1rill) [inline profile/src/main/java/org/openedx/profile/presentation/edit/EditProfileViewModel.kt:162]: nit: please improve formatting

File: profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt [hunk 11]
```diff
@@ -134,12 +137,18 @@ class ProfileViewModel(
     }
 
     fun logout() {
+        logProfileEvent(ProfileAnalyticEvent.LOGOUT_CLICKED.event, buildMap {
```
Comment (k1rill) [inline profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt:140]: nit: please improve formatting

File: profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt [hunk 12]
```diff
@@ -134,12 +137,18 @@ class ProfileViewModel(
...
+            put(ProfileAnalyticKey.NAME.key, ProfileAnalyticValue.LOGOUT_CLICKED.biValue)
+        })
         viewModelScope.launch {
             try {
                 workerController.removeModels()
                 withContext(dispatcher) {
                     interactor.logout()
                 }
+                logProfileEvent(ProfileAnalyticEvent.LOGGED_OUT.event, buildMap {
```
Comment (k1rill) [inline profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt:149]: nit: please improve formatting

File: profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt [hunk 13]
```diff
@@ -171,12 +179,16 @@ class ProfileViewModel(
                 data.account
             )
         }
-        analytics.profileEditClickedEvent()
+        logProfileEvent(ProfileAnalyticEvent.EDIT_CLICKED.event, buildMap {
```
Comment (k1rill) [inline profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt:182]: nit: please improve formatting

File: profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt [hunk 14]
```diff
@@ -171,12 +179,16 @@ class ProfileViewModel(
...
+            put(ProfileAnalyticKey.NAME.key, ProfileAnalyticValue.EDIT_CLICKED.biValue)
+        })
     }
 
     fun profileVideoSettingsClicked(fragmentManager: FragmentManager) {
         router.navigateToVideoSettings(fragmentManager)
-        analytics.profileVideoSettingsClickedEvent()
+        logProfileEvent(ProfileAnalyticEvent.VIDEO_SETTING_CLICKED.event, buildMap {
```
Comment (k1rill) [inline profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt:189]: nit: please improve formatting

File: profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt [hunk 15]
```diff
@@ -185,7 +197,9 @@ class ProfileViewModel(
             title = resourceManager.getString(R.string.core_privacy_policy),
             url = configuration.agreementUrls.privacyPolicyUrl,
         )
-        analytics.privacyPolicyClickedEvent()
+        logProfileEvent(ProfileAnalyticEvent.PRIVACY_POLICY_CLICKED.event, buildMap {
```
Comment (k1rill) [inline profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt:200]: nit: please improve formatting

File: profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt [hunk 16]
```diff
@@ -194,7 +208,9 @@ class ProfileViewModel(
             title = resourceManager.getString(R.string.core_cookie_policy),
             url = configuration.agreementUrls.cookiePolicyUrl,
         )
-        analytics.cookiePolicyClickedEvent()
+        logProfileEvent(ProfileAnalyticEvent.COOKIE_POLICY_CLICKED.event, buildMap {
```
Comment (k1rill) [inline profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt:211]: nit: please improve formatting

File: profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt [hunk 17]
```diff
@@ -203,11 +219,15 @@ class ProfileViewModel(
             title = resourceManager.getString(R.string.core_data_sell),
             url = configuration.agreementUrls.dataSellConsentUrl,
         )
-        analytics.dataSellClickedEvent()
+        logProfileEvent(ProfileAnalyticEvent.DATA_SELL_CLICKED.event, buildMap {
```
Comment (k1rill) [inline profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt:222]: nit: please improve formatting

File: profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt [hunk 18]
```diff
@@ -203,11 +219,15 @@ class ProfileViewModel(
...
+            put(ProfileAnalyticKey.NAME.key, ProfileAnalyticValue.DATA_SELL_CLICKED.biValue)
+        })
     }
 
     fun faqClicked() {
-        analytics.faqClickedEvent()
+        logProfileEvent(ProfileAnalyticEvent.FAQ_CLICKED.event, buildMap {
```
Comment (k1rill) [inline profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt:228]: nit: please improve formatting

File: profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt [hunk 19]
```diff
@@ -216,7 +236,9 @@ class ProfileViewModel(
             title = resourceManager.getString(R.string.core_terms_of_use),
             url = configuration.agreementUrls.tosUrl,
         )
-        analytics.termsOfUseClickedEvent()
+        logProfileEvent(ProfileAnalyticEvent.TERMS_OF_USE_CLICKED.event, buildMap {
```
Comment (k1rill) [inline profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt:239]: nit: please improve formatting

File: profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt [hunk 20]
```diff
@@ -225,7 +247,9 @@ class ProfileViewModel(
             feedbackEmailAddress = config.getFeedbackEmailAddress(),
             appVersion = appData.versionName
         )
-        analytics.emailSupportClickedEvent()
+        logProfileEvent(ProfileAnalyticEvent.CONTACT_SUPPORT_CLICKED.event, buildMap {
```
Comment (k1rill) [inline profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt:250]: nit: please improve formatting

File: profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt [hunk 21]
```diff
@@ -239,4 +263,10 @@ class ProfileViewModel(
         )
     }
 
+    private fun logProfileEvent(event: String, params: Map<String, Any?>) {
+        analytics.logEvent(event, buildMap {
```
Comment (k1rill) [inline profile/src/main/java/org/openedx/profile/presentation/profile/ProfileViewModel.kt:267]: nit: please improve formatting

**Review by k1rill — COMMENTED** [review]
Thank you for the very accurate job! Just minor nits

File: auth/src/main/java/org/openedx/auth/presentation/signin/SignInViewModel.kt (hunk 4 above)
Comment (omerhabib26) [reply auth/src/main/java/org/openedx/auth/presentation/signin/SignInViewModel.kt:180]: removed

File: course/src/main/java/org/openedx/course/presentation/container/CourseContainerFragment.kt (hunk 7 above)
Comment (omerhabib26) [reply course/src/main/java/org/openedx/course/presentation/container/CourseContainerFragment.kt:247]: replaced with `else -> {}`

**Review by k1rill — APPROVED** [review]
LGTM

File: core/src/main/java/org/openedx/core/presentation/CoreAnalytics.kt [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+        "edx.bi.app.videos.download.toggle"
+    ),
+    VIDEO_DOWNLOAD_SUBSECTION(
+        "Video:Download Subsection",
+        "edx.bi.video.download.subsection"
+    ),
+    VIDEO_DELETE_SUBSECTION(
+        "Videos:Delete Subsection",
+        "edx.bi.app.video.delete.subsection"
+    ),
```
Comment (HamzaIsrar12) [inline core/src/main/java/org/openedx/core/presentation/CoreAnalytics.kt:27]: These events are missing these attributes:
- category: videos,
- number_of_videos: count

File: course/src/main/java/org/openedx/course/presentation/outline/CourseOutlineViewModel.kt [hunk 23]
```diff
@@ -124,6 +134,7 @@ class CourseOutlineViewModel(
                     UIMessage.ToastMessage(resourceManager.getString(courseR.string.course_can_download_only_with_wifi))
             }
         } else {
+            logSubsectionDownloadEvent(id)
```
Comment (HamzaIsrar12) [inline course/src/main/java/org/openedx/course/presentation/outline/CourseOutlineViewModel.kt:137]: It's not getting called. 🤐

File: course/src/main/java/org/openedx/course/presentation/unit/video/BaseVideoViewModel.kt
Comment (HamzaIsrar12) [inline course/src/main/java/org/openedx/course/presentation/unit/video/BaseVideoViewModel.kt:1]: The `currentVideoTime` is always returning 0 in evens.

**Review by HamzaIsrar12 — CHANGES_REQUESTED** [review]
Some things to look into:

- Discovery Events are implemented for Native experience only.
- CalendarSynnc Action Events is triggering for all cases.
- Please follow a similar convention in Analytics Screen & Keys enum names.

File: course/src/main/java/org/openedx/course/presentation/detail/CourseDetailsViewModel.kt [hunk 24]
```diff
@@ -127,15 +129,30 @@ class CourseDetailsViewModel(
...
+        event: CourseAnalyticsEvent,
+        courseId: String, courseTitle: String,
+    ) {
+        analytics.logEvent(
+            event.eventName,
+            buildMap {
+                put(CourseAnalyticKey.NAME.key, event.biValue)
+                put(CourseAnalyticKey.COURSE_ID.key, courseId)
+                put(CourseAnalyticKey.COURSE_NAME.key, courseTitle)
+                put(CourseAnalyticKey.CONVERSION.key, courseId)
```
[the remark this answers is not in the document]
Comment (omerhabib26) [reply course/src/main/java/org/openedx/course/presentation/detail/CourseDetailsViewModel.kt:154]: don't know exactly, already implemented in Market app.

File: course/src/main/java/org/openedx/course/presentation/unit/video/BaseVideoViewModel.kt
Comment (omerhabib26) [reply course/src/main/java/org/openedx/course/presentation/unit/video/BaseVideoViewModel.kt:1]: we can create a bug ticket to tackle currentVideoTime as according to current implementation it is hard to trace why its not working

**Review by HamzaIsrar12 — CHANGES_REQUESTED** [review]
- CalendarSync isn't creating events on Calendar
- Discovery Events are implemented for Native experience only.
- WhatsNew:Close event is not implemented

**Review by HamzaIsrar12 — APPROVED** [review]
Thank you, Moin, great work !! 🏎️

File: course/src/test/java/org/openedx/course/presentation/container/CourseContainerViewModelTest.kt [hunk 25]
```diff
@@ -184,17 +192,19 @@ class CourseContainerViewModelTest {
...
         )
         every { networkConnection.isOnline() } returns true
         coEvery { interactor.preloadCourseStructure(any()) } returns Unit
         every { interactor.getCourseStructureFromCache() } returns courseStructure
+        every { analytics.logEvent(any(), any()) } returns Unit
         viewModel.preloadCourseStructure()
         advanceUntilIdle()
 
         coVerify(exactly = 1) { interactor.preloadCourseStructure(any()) }
+        verify(exactly = 1) { analytics.logEvent(any(), any()) }
```
Comment (volodymyr-chekyrta) [inline course/src/test/java/org/openedx/course/presentation/container/CourseContainerViewModelTest.kt:207]: Should check specific value, not `any`

File: course/src/test/java/org/openedx/course/presentation/container/CourseContainerViewModelTest.kt [hunk 26]
```diff
@@ -128,16 +130,18 @@ class CourseContainerViewModelTest {
...
+            analytics,
         )
         every { networkConnection.isOnline() } returns true
         coEvery { interactor.preloadCourseStructure(any()) } throws UnknownHostException()
+        every { analytics.logEvent(any(), any()) } returns Unit
         viewModel.preloadCourseStructure()
         advanceUntilIdle()
 
         coVerify(exactly = 1) { interactor.preloadCourseStructure(any()) }
+        verify(exactly = 1) { analytics.logEvent(any(), any()) }
```
Comment (volodymyr-chekyrta) [inline course/src/test/java/org/openedx/course/presentation/container/CourseContainerViewModelTest.kt:144]: Should check specific value, not `any`

File: course/src/test/java/org/openedx/course/presentation/detail/CourseDetailsViewModelTest.kt [hunk 27]
```diff
@@ -271,21 +267,19 @@ class CourseDetailsViewModelTest {
...
 
         delay(200)
-        viewModel.enrollInACourse("")
+        viewModel.enrollInACourse("", "")
         advanceUntilIdle()
 
         coVerify(exactly = 1) { interactor.enrollInACourse(any()) }
-        verify(exactly = 1) { analytics.courseEnrollClickedEvent(any(), any()) }
-        verify(exactly = 1) { analytics.courseEnrollSuccessEvent(any(), any()) }
+        verify(exactly = 2) { analytics.logEvent(any(), any()) }
```
Comment (volodymyr-chekyrta) [inline course/src/test/java/org/openedx/course/presentation/detail/CourseDetailsViewModelTest.kt:282]: Should check specific value, not `any`

File: course/src/test/java/org/openedx/course/presentation/outline/CourseOutlineViewModelTest.kt [hunk 28]
```diff
@@ -445,12 +454,14 @@ class CourseOutlineViewModelTest {
...
             preferencesManager,
             analytics,
+            coreAnalytics,
             downloadDao,
             workerController
         )
 
         viewModel.saveDownloadModels("", "")
         advanceUntilIdle()
+        verify(exactly = 1) { coreAnalytics.logEvent(any(), any()) }
```
Comment (volodymyr-chekyrta) [inline course/src/test/java/org/openedx/course/presentation/outline/CourseOutlineViewModelTest.kt:464]: Should check specific value, not `any`

File: course/src/test/java/org/openedx/course/presentation/unit/video/VideoUnitViewModelTest.kt [hunk 29]
```diff
@@ -70,6 +76,7 @@ class VideoUnitViewModelTest {
                 any()
             )
         }
+        verify(exactly = 1) { courseAnalytics.logEvent(any(), any()) }
```
Comment (volodymyr-chekyrta) [inline course/src/test/java/org/openedx/course/presentation/unit/video/VideoUnitViewModelTest.kt:79]: Should check specific value, not `any`

File: course/src/test/java/org/openedx/course/presentation/unit/video/VideoUnitViewModelTest.kt [hunk 30]
```diff
@@ -96,6 +105,7 @@ class VideoUnitViewModelTest {
...
```
Comment (volodymyr-chekyrta) [inline course/src/test/java/org/openedx/course/presentation/unit/video/VideoUnitViewModelTest.kt:108]: Should check specific value, not `any`

File: course/src/test/java/org/openedx/course/presentation/unit/video/VideoViewModelTest.kt [hunk 31]
```diff
@@ -67,18 +72,21 @@ class VideoViewModelTest {
                 any()
             )
         }
+        verify(exactly = 1) { courseAnalytics.logEvent(any(), any()) }
```
Comment (volodymyr-chekyrta) [inline course/src/test/java/org/openedx/course/presentation/unit/video/VideoViewModelTest.kt:75]: Should check specific value, not `any`

File: course/src/test/java/org/openedx/course/presentation/unit/video/VideoViewModelTest.kt [hunk 32]
```diff
@@ -87,6 +95,7 @@ class VideoViewModelTest {
...
```
Comment (volodymyr-chekyrta) [inline course/src/test/java/org/openedx/course/presentation/unit/video/VideoViewModelTest.kt:98]: Should check specific value, not `any`

File: auth/src/main/java/org/openedx/auth/presentation/AuthAnalytics.kt [hunk 33]
```diff
@@ -1,11 +1,60 @@
...
-    fun userLoginEvent(method: String)
-    fun signUpClickedEvent()
-    fun createAccountClickedEvent(provider: String)
-    fun registrationSuccessEvent(provider: String)
-    fun forgotPasswordClickedEvent()
-    fun resetPasswordClickedEvent(success: Boolean)
     fun setUserIdForSession(userId: Long)
-}
\ No newline at end of file
+    fun logEvent(event: String, params: Map<String, Any?>)
```
[the remark this answers is not in the document]
Comment (omerhabib26) [reply auth/src/main/java/org/openedx/auth/presentation/AuthAnalytics.kt:5]: > What is the purpose of removing these methods?
* We're moving methods to ViewModels for better organization and maintenance. Each ViewModel handles its own analytics, making future changes easier without affecting other code. This promotes code reuse and reduces the risk of errors.

> I see a similar approach in a few other interfaces, but that leads to inconsistency with CourseAnalytics, DashboardAnalytics, DiscoveryAnalytics, DiscussionAnalytics, etc.
- Some methods may still exist from a previous approach, but lack proper event details. These can be addressed in an improvement ticket, ensuring that any remaining methods are properly handled and documented for clarity and future maintenance.

> We should select one approach and follow it for all modules. My +1 for keeping methods in interfaces: They help quickly understand what this interface does, same to table of contents in books;
abstract method logEvent cannot provide this information.
- In some cases, we use default logEvent instead custom method which can be improved in a separate ticket for better understanding and readability.

If you have any alternative solutions in mind, we're open to considering them as well.

File: course/src/test/java/org/openedx/course/presentation/container/CourseContainerViewModelTest.kt [hunk 34]
```diff
@@ -156,16 +161,18 @@ class CourseContainerViewModelTest {
...
+            analytics,
         )
         every { networkConnection.isOnline() } returns true
         coEvery { interactor.preloadCourseStructure(any()) } throws Exception()
+        every { analytics.logEvent(any(), any()) } returns Unit
         viewModel.preloadCourseStructure()
         advanceUntilIdle()
 
         coVerify(exactly = 1) { interactor.preloadCourseStructure(any()) }
+        verify(exactly = 1) { analytics.logEvent(any(), any()) }
```
[the remark this answers is not in the document]
Comment (omerhabib26) [reply course/src/test/java/org/openedx/course/presentation/container/CourseContainerViewModelTest.kt:175]: Hey @volodymyr-chekyrta, 
As per the current implementation, we are not verifying the method with a specific value.
i.e, `coVerify(exactly = 1) { interactor.preloadCourseStructure(any()) }`

[the remark this answers is not in the document]
Comment (volodymyr-chekyrta) [reply course/src/test/java/org/openedx/course/presentation/container/CourseContainerViewModelTest.kt:175]: In the case of `interactor.preloadCourseStructure(any())`, it's not important what value is passed as an argument, but in the case of `analytics.logEvent`, we need to check what event we logged.

[the remark this answers is not in the document]
Comment (omerhabib26) [reply course/src/test/java/org/openedx/course/presentation/container/CourseContainerViewModelTest.kt:175]: I have create a new PR for these changes Please have a look

**Pull request merged**